src/EventListener/ProductTranslationUpdateListener.php

```diff
@@ -0,0 +1,139 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Setono\SyliusRedirectPlugin\EventListener;
+
+use Doctrine\ORM\Event\LifecycleEventArgs;
+use Setono\SyliusRedirectPlugin\Model\RedirectInterface;
+use Setono\SyliusRedirectPlugin\Resolver\InfiniteLoopResolverInterface;
+use Sylius\Component\Core\Model\ProductTranslationInterface;
+use Sylius\Component\Resource\Exception\UpdateHandlingException;
+use Sylius\Component\Resource\Factory\FactoryInterface;
+use Sylius\Component\Resource\Repository\RepositoryInterface;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\HttpFoundation\Session\Flash\FlashBagInterface;
+use Symfony\Component\Routing\RouterInterface;
+use Symfony\Component\Validator\ConstraintViolationInterface;
+use Symfony\Component\Validator\Validator\ValidatorInterface;
+
+final class ProductTranslationUpdateListener
+{
+    /**
+     * @var \Symfony\Component\HttpFoundation\Request|null
+     */
+    private $request;
+    /**
+     * @var FactoryInterface
+     */
+    private $redirectionFactory;
+    /**
+     * @var RepositoryInterface
+     */
+    private $redirectionRepository;
+    /**
+     * @var InfiniteLoopResolverInterface
+     */
+    private $infiniteLoopResolver;
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+    /**
+     * @var ValidatorInterface
+     */
+    private $validator;
+    /**
+     * @var FlashBagInterface
+     */
+    private $flashBag;
+
+    public function __construct(RequestStack $requestStack,
+                                FactoryInterface $redirectionFactory,
+                                RepositoryInterface $redirectionRepository,
+                                InfiniteLoopResolverInterface $infiniteLoopResolver,
+                                RouterInterface $router,
+                                ValidatorInterface $validator,
+                                FlashBagInterface $flashBag
+    )
+    {
+        $this->request = $requestStack->getCurrentRequest();
+        $this->redirectionFactory = $redirectionFactory;
+        $this->redirectionRepository = $redirectionRepository;
+        $this->infiniteLoopResolver = $infiniteLoopResolver;
+        $this->router = $router;
+        $this->validator = $validator;
+        $this->flashBag = $flashBag;
+    }
+
+    public function postUpdate(LifecycleEventArgs $event): void
```

Comment:
And shouldn't it be preUpdate you listen to?

Reply:
Using a preUpdate listener is quite tricky since persisting new entities is not possible. They are simply ignored. I could use onFlush, but postUpdate is really cleaner in terms of code.

Reply:
FWIW, if you need to persist a new entity, you should always do it in `preFlush` or `onFlush`: there is no guarantee for the state of the UnitOfWork after that. As you've noticed, it doesn't persist entities you manage later, and calling `onFlush` in a listener is a big no-no.

Reply:
Thank you for your review, and thanks for the slack discussion. I believe that now the changes are made early enough to be sure that changes are persisted to DB.